tools/rpc-client/RpcClient.hs

```diff
@@ -22,81 +27,125 @@ import Data.Aeson.KeyMap qualified as JsonKeyMap
 import Data.Bifunctor
 import Data.ByteString.Lazy.Char8 qualified as BS
 import Data.Char (isDigit)
+import Data.Int (Int64)
 import Data.List.Extra
-import Data.Maybe (isNothing)
+import Data.Maybe (isNothing, mapMaybe)
 import Data.Text qualified as Text
 import Data.Vector as Array (fromList)
+import GHC.IO.Exception (IOErrorType (..), IOException (..))
 import Network.Run.TCP
 import Network.Socket
 import Network.Socket.ByteString.Lazy
 import Options.Applicative
 import System.Clock
 import System.Directory
 import System.Exit
+import System.FilePath
 import System.IO
+import System.IO.Extra
 import System.Process
 
 import Booster.JsonRpc (rpcJsonConfig)
+import Booster.JsonRpc.Utils (diffJson, isIdentical, renderResult)
 import Booster.Syntax.Json qualified as Syntax
 
 import Debug.Trace
 
 main :: IO ()
 main = do
-    Options{host, port, mode, optionFile, options, postProcessing, prettify, time, dryRun} <-
-        execParser parseOptions
-    request <-
-        trace "[Info] Preparing request data" $
-            prepareRequestData mode optionFile options
-    when dryRun $ do
-        traceM "[Info] Dry-run mode, just showing request instead of sending"
-        let write
-                | Just (Expect True file) <- postProcessing =
-                    trace ("[Info] Writing request to file " <> file) (BS.writeFile file)
-                | otherwise = BS.putStrLn
-            reformat = Json.encodePretty' rpcJsonConfig . Json.decode @Json.Value
-        write $ if not prettify then request else reformat request
-        exitSuccess
-    runTCPClient host (show port) $ \s -> do
-        start <- getTime Monotonic
-        trace "[Info] Sending request..." $
-            sendAll s request
-        response <- readResponse s 8192
-        end <- getTime Monotonic
-        let modeFile = getModeFile mode
-            timeStr = timeSpecs start end
-        hPutStrLn stderr $ "[info] Round trip time for request '" <> modeFile <> "' was " <> timeStr
-        when time $ do
-            hPutStrLn stderr $ "[Info] Saving timing for " <> modeFile
-            writeFile (modeFile <> ".time") timeStr
-
-        trace "[Info] Response received." $
-            postProcess prettify postProcessing response
-        shutdown s ShutdownReceive
+    Options{common, runOptions} <- execParser parseOptions
+    case runOptions of
+        RunTarball tarFile keepGoing ->
+            runTarball common.host common.port tarFile keepGoing
+        RunSingle mode optionFile options processingOptions -> do
+            let ProcessingOptions{postProcessing, prettify, time, dryRun} = processingOptions
+
+            request <-
+                trace "[Info] Preparing request data" $
+                    prepareRequestData mode optionFile options
```

Comment:
This is still using `trace` throughout but we should convert it to using a logger (just `MonadLoggerIO` like the server, I would guess)
Also, we should send a `Cancel` for the request in-flight when the program is interrupted (`^C`).

I left those two changes for a follow-up PR, as this one is certainly enough change for one round of review.